### PR TITLE
Shortened block inside sync

### PR DESCRIPTION
**Pull request by NIkonDSL**

### CODE REVIEW

File: serializer-base64/src/main/java/com/vlkan/hrrs/serializer/base64/Base64HttpRequestRecordWriter.java [hunk 1]
```diff
@@ -5,45 +5,58 @@
...
-
-    public Base64HttpRequestRecordWriter(HttpRequestRecordWriterTarget<String> target, Base64Encoder encoder) {
+    protected Base64HttpRequestRecordWriter(HttpRequestRecordWriterTarget<String> target, Base64Encoder encoder) {
         this.target = checkNotNull(target, "target");
         this.encoder = checkNotNull(encoder, "encoder");
     }
 
+    public static Base64HttpRequestRecordWriter createBase64HttpRequestRecordWriter(HttpRequestRecordWriterTarget<String> target, Base64Encoder encoder) {
+        if (System.getProperty("use.hrrs.queue") != null) {
+            return new QueueBase64HttpRequestRecordWriter(target, encoder);
```
Comment (vy) [inline serializer-base64/src/main/java/com/vlkan/hrrs/serializer/base64/Base64HttpRequestRecordWriter.java:28]: I cannot see who provides `QueueBase64HttpRequestRecordWriter`. IIRC, it is not included in `master`.

File: serializer-base64/src/main/java/com/vlkan/hrrs/serializer/base64/Base64HttpRequestRecordWriter.java [hunk 2]
```diff
@@ -5,45 +5,58 @@
...
-                target.write(FIELD_SEPARATOR);
-                target.write(record.getGroupName());
-                target.write(FIELD_SEPARATOR);
-                target.write(record.getMethod().toString());
-                target.write(FIELD_SEPARATOR);
-                byte[] recordBytes = writeRecord(record);
-                String encodedRecordBytes = encoder.encode(recordBytes);
-                target.write(encodedRecordBytes);
-                target.write(RECORD_SEPARATOR);
+                target.write(toBeRecorded.toString());
```
Comment (vy) [inline serializer-base64/src/main/java/com/vlkan/hrrs/serializer/base64/Base64HttpRequestRecordWriter.java:57]: Indeed this commit shortens the _LoC_ in the `synchronized` block, though I am not convinced that this performs better. You replace the lengthy `synchronized` block with a one-liner. Though the amount of bytes need be written to the `target` are exactly the same, this new scheme only avoids the time spent in `encode(byte[])`. Further, one also need to pay the cost of extra `StringBuilder` allocation each time. This allocation cost can be avoided using an object pool, but that is another story. In conclusion, do you have empirical evidence to prove that this approach performs better? Did it cause a bottleneck in your case?

**Comment by NIkonDSL** [comment]
I've taken your library to myself and you may find missing file there.
I shortened time spending in sync block in original file and Queue version
has no sync at all due to internal queue, where all work is done in
consumer thread, producers only provide info for processing. It works worse
for the test where you've create/remove Writer/reader all the time, but
works faster for common case where Writer is single

пн, 1 бер. 2021 о 11:23 Volkan Yazıcı <notifications@github.com> пише:

**Comment by vy** [comment]
@NIkonDSL, in a queue setup, writer indeed avoids the synchronization cost, though this time the call-site need to synchronize for the queue access. Would you mind elaborating a bit on your use case, please? Did you have a performance issue? What was the motivation for async writes? What do you use HRRS for?

**Pull request merged**